File: code/modM_MIIDonly.R
# IID spatial effects only model for M_M (MICS-only) data
# Uses same integration point structure as the full models
# Replaces BYM2 with IID spatial random effects + cluster-level nugget

fitMM_IIDonly = function(datDHS=ed, datMICS=edMICS, inputsMDM=NULL,
                         intPtsMICS=NULL, intPtsDHS=NULL,
                         KMICS=100,
                         KDHSurb = 11,
                         JInnerUrban = 3,
                         KDHSrur = 16,
                         JInnerRural = 3,
                         JOuterRural = 1, admMICS=admFinal, adm2DHS=adm2Full,
                         alpha_pri = c(0, 100^2),
                         beta_pri = c(0, sqrt(1000)),
                         pc.iidPrec=list(u=1, alpha=0.5),
                         pc.expPrec=list(u=1, alpha=0.5),
                         covariates=NULL,
                         fixedEffectsOnly=FALSE,
                         maxit=1000, adm2AsCovariate=FALSE,
                         tolSeq = c(1e-06), getSDs=TRUE, doMCMC=FALSE) {
  
  # make sure Stratum variable exists in MICS data
  if(!("Stratum" %in% names(datMICS))) {
    datMICS$Stratum = adm2ToStratumMICS(datMICS$subarea)
  }
  
  # make sure all the names in our datasets we expect to be there are
  nameTab = rbind(c("N", "ns"),
                  c("N", "n"),
                  c("Z", "ys"),
                  c("Z", "y"))
  for(i in 1:nrow(nameTab)) {
    theseNames = nameTab[i,]
    fromN = theseNames[1]
    toN = theseNames[2]
    if(!(toN %in% names(datMICS))) {
      datMICS[[toN]] = datMICS[[fromN]]
    }
    if(!(toN %in% names(datDHS))) {
      datDHS[[toN]] = datDHS[[fromN]]
    }
  }
  
  # generate all necessary inputs if need be
  if(is.null(inputsMDM)) {
    print("Making M_DM inputs for IID model...")
    
    inputsMDM = makeInputsMDM(datDHS, datMICS,
                              intPtsMICS=intPtsMICS, intPtsDHS=intPtsDHS,
                              KMICS=KMICS,
                              KDHSurb = KDHSurb,
                              JInnerUrban = JInnerUrban,
                              KDHSrur = KDHSrur,
                              JInnerRural = JInnerRural,
                              JOuterRural = JOuterRural,
                              admMICS=admMICS, adm2DHS=adm2DHS,
                              adm2AsCovariate=adm2AsCovariate)
  }
  
  thisEnv = environment()
  list2env(inputsMDM, envir=thisEnv)
  
  datMICS = sortByCol(datMICS, "Stratum", admMICS$NAME_FINAL)
  
  # set priors ----
  lambdaTau = getLambdaPCprec(u=pc.iidPrec$u, alpha=pc.iidPrec$alpha)
  lambdaTauEps = getLambdaPCprec(u=pc.expPrec$u, alpha=pc.expPrec$alpha)
  
  # number of areas from area indices (0-indexed)
  nAreas = max(c(areaidxlocUrbanMICS, areaidxlocRuralMICS)) + 1
  
  # Subset covariates if requested ----
  allCovNames = colnames(intPtsMICS$XUrb)
  if(!is.null(covariates)) {
    keepIdx = which(allCovNames %in% covariates)
    cat("  Subsetting covariates to:", allCovNames[keepIdx], "\n")
    intPtsMICS$XUrb = intPtsMICS$XUrb[, keepIdx, drop=FALSE]
    intPtsMICS$XRur = intPtsMICS$XRur[, keepIdx, drop=FALSE]
  } else {
    cat("  Using all covariates:", allCovNames, "\n")
  }
  
  # Specify inputs for TMB ----
  if(fixedEffectsOnly) {
    rand_effs <- c('nuggetUrbMICS', 'nuggetRurMICS')
    mapList <- list(u_spatial = factor(rep(NA, nAreas)),
                    log_tau   = factor(NA))
  } else {
    rand_effs <- c('alpha', 'beta', 'u_spatial', 'nuggetUrbMICS', 'nuggetRurMICS')
    mapList <- list()
  }
  
  data_full = list(
    y_iUrbanMICS=ysUrbMICS,
    y_iRuralMICS=ysRurMICS,
    n_iUrbanMICS=nsUrbMICS,
    n_iRuralMICS=nsRurMICS,
    areaidxlocUrbanMICS=areaidxlocUrbanMICS,
    areaidxlocRuralMICS=areaidxlocRuralMICS,
    X_betaUrbanMICS=intPtsMICS$XUrb,
    X_betaRuralMICS=intPtsMICS$XRur,
    wUrbanMICS=intPtsMICS$wUrban,
    wRuralMICS=intPtsMICS$wRural,
    nAreas=nAreas,
    alpha_pri=alpha_pri,
    beta_pri=beta_pri,
    lambdaTau=lambdaTau,
    lambdaTauEps=lambdaTauEps
  )
  
  # initial parameters
  initUrbP = sum(c(data_full$y_iUrbanMICS))/sum(c(data_full$n_iUrbanMICS))
  initRurP = sum(c(data_full$y_iRuralMICS))/sum(c(data_full$n_iRuralMICS))
  initAlpha = logit(initRurP)
  initBeta1 = logit(initUrbP) - initAlpha
  
  tmb_params <- list(log_tau = 0,
                     log_tauEps = 0,
                     alpha = initAlpha,
                     beta = c(initBeta1, rep(0, ncol(data_full$X_betaUrbanMICS)-1)),
                     u_spatial = rep(0, nAreas),
                     nuggetUrbMICS = rep(0, length(data_full$y_iUrbanMICS)),
                     nuggetRurMICS = rep(0, length(data_full$y_iRuralMICS)))
  
  # make TMB fun and grad ----
  unloadDynlibs()
  
  if(!any(file.exists(paste0("code/modM_MIIDonly", c(".o", ".so", ".dll"))))) {
    print("compiling code/modM_MIIDonly.cpp...")
    compile("code/modM_MIIDonly.cpp", framework="TMBad", safebounds=FALSE)
  }
  
  dyn.load(dynlib("code/modM_MIIDonly"))
  TMB::config(tmbad.sparse_hessian_compress = 1)
  obj <- MakeADFun(data=data_full,
                   parameters=tmb_params,
                   random=rand_effs,
                   map=mapList,
                   hessian=TRUE,
                   DLL='modM_MIIDonly')
  
  lower = rep(-10, length(obj[['par']]))
  upper = rep( 10, length(obj[['par']]))
  
  # wrapper functions that print out parameters and function values
  funWrapper = function(par, badParVal=1e10) {
    if(any(par < lower) || any(par > upper)) {
      return(badParVal)
    }
    objVal = obj[['fn']](par)
    print(paste0("objective: ", objVal, " for log_tau=", par))
    objVal
  }
  
  grWrapper = function(par, badParVal=1e10) {
    if(any(par < lower) || any(par > upper)) {
      return(rep(badParVal, length(par)))
    }
    obj[['gr']](par)
  }
  
  # Run TMB ----
  if(!doMCMC) {
    optPar = obj$par
    startTime = proc.time()[3]
    for(thisTol in tolSeq) {
      obj$env$inner.control = list(maxit=maxit, tol10=thisTol)
      obj$env$tracepar = TRUE
      print(paste0("optimizing for tol = ", thisTol, "."))
      opt1 <- optim(par=optPar, fn=funWrapper, gr=grWrapper,
                    method = c("BFGS"), hessian = FALSE, control=list(reltol=thisTol))
      optPar = opt1$par
      if(!is.null(opt1$message)) {
        print(paste0("error for tol = ", thisTol, ". Message:"))
        print(opt1$message)
        next
      } else {
        print(paste0("completed optimization for tol = ", thisTol, ""))
        
        if(getSDs) {
          print("getting standard errors...")
          sdTime = system.time(
            SD0 <- TMB::sdreport(obj, getJointPrecision=TRUE,
                                 bias.correct = TRUE,
                                 bias.correct.control = list(sd = TRUE))
          )[3]
          print(sdTime/60)
          
          if(SD0$pdHess) {
            print("Optimization and PD hess calculation done!")
            break
          } else {
            print("Hessian not PD. Rerunning optimization with stricter tol...")
          }
        } else {
          print("getting report (no standard errors)...")
          sdTime = system.time(
            SD0 <- obj$report()
          )[3]
        }
      }
    }
    endTime = proc.time()[3]
    totalTime = endTime - startTime
    print(paste0("optimization took ", totalTime/60, " minutes"))
  } else {
    startTime = proc.time()[3]
    fit <- tmbstan(obj=obj, silent=FALSE, laplace=FALSE,
                   iter=4000, warmup=2000, chains=1)
    endTime = proc.time()[3]
    totalTime = endTime - startTime
    sdTime = 0
    SD0 = fit
    print(paste0("MCMC took ", totalTime/60, " minutes"))
  }
  
  list(TMBobj=obj, TMBsd=SD0, totalTime=totalTime, sdTime=sdTime)
}

File: code/modM_MIIDonly.cpp
// IID spatial effects only model - simplified for testing
// M_M (MICS-only) data
// Replaces BYM2 with IID spatial random effects + cluster-level nugget
// Uses same integration point structure as the full models

#include <TMB.hpp>
#include <Eigen/Sparse>
#include <vector>
using namespace density;
using Eigen::SparseMatrix;

// PC prior on precision parameter (on log scale)
template<class Type>
Type dPCPriTau(Type logTau, Type lambda)
{
  Type tau = exp(logTau);
  Type ldensity = log(lambda) - log(2.0) - Type(1.5)*logTau - lambda/sqrt(tau);
  Type ljacobian = logTau;
  return ldensity + ljacobian;
}

template<class Type>
Type objective_function<Type>::operator() ()
{
  // Data
  DATA_VECTOR( y_iUrbanMICS );
  DATA_VECTOR( y_iRuralMICS );
  DATA_VECTOR( n_iUrbanMICS );
  DATA_VECTOR( n_iRuralMICS );
  DATA_IVECTOR(areaidxlocUrbanMICS);
  DATA_IVECTOR(areaidxlocRuralMICS);
  DATA_MATRIX( X_betaUrbanMICS );
  DATA_MATRIX( X_betaRuralMICS );
  DATA_ARRAY( wUrbanMICS );
  DATA_ARRAY( wRuralMICS );
  DATA_INTEGER( nAreas );
  
  // Prior parameters
  DATA_VECTOR( alpha_pri );
  DATA_VECTOR( beta_pri );
  DATA_SCALAR( lambdaTau );
  DATA_SCALAR( lambdaTauEps );
  
  // Fixed effects
  PARAMETER( log_tau );
  PARAMETER( log_tauEps );
  
  // Intercept
  PARAMETER( alpha );
  PARAMETER_VECTOR( beta );
  
  // IID spatial random effects
  PARAMETER_VECTOR( u_spatial );
  
  // Cluster-level nugget effects
  PARAMETER_VECTOR( nuggetUrbMICS );
  PARAMETER_VECTOR( nuggetRurMICS );
  
  // Dimensions
  int num_iUrbanMICS = y_iUrbanMICS.size();
  int num_iRuralMICS = y_iRuralMICS.size();
  int n_integrationPointsUrbanMICS = wUrbanMICS.cols();
  int n_integrationPointsRuralMICS = wRuralMICS.cols();
  
  // Transform
  Type tau = exp(log_tau);
  Type sigma_u = exp(Type(-0.5) * log_tau);
  Type sigmaEps = exp(Type(-0.5) * log_tauEps);
  
  Type jnll = 0.0;
  
  // (1) IID spatial random effects prior
  for(int i = 0; i < u_spatial.size(); i++) {
    jnll -= dnorm(u_spatial(i), Type(0), sigma_u, true);
  }
  
  // (2) Priors
  // PC prior for tau (IID spatial)
  jnll -= dPCPriTau(log_tau, lambdaTau);
  
  // PC prior for tauEps (nugget)
  jnll -= dPCPriTau(log_tauEps, lambdaTauEps);
  
  // Nugget priors
  for(int i = 0; i < nuggetUrbMICS.size(); i++) {
    jnll -= dnorm(nuggetUrbMICS(i), Type(0), sigmaEps, true);
  }
  for(int i = 0; i < nuggetRurMICS.size(); i++) {
    jnll -= dnorm(nuggetRurMICS(i), Type(0), sigmaEps, true);
  }
  
  // Prior for intercept
  jnll -= dnorm(alpha, alpha_pri(0), alpha_pri(1), true);
  
  // Prior for covariates
  for(int i = 0; i < beta.size(); i++) {
    jnll -= dnorm(beta(i), beta_pri(0), beta_pri(1), true);
  }
  
  // (3) Likelihood with integration points (same structure as Marg model)
  vector<Type> fe_iUrbanMICS = X_betaUrbanMICS * beta;
  vector<Type> fe_iRuralMICS = X_betaRuralMICS * beta;
  
  int thisIndex;
  Type thisWeight;
  Type thislik;
  
  // Urban MICS
  for (int obsI = 0; obsI < num_iUrbanMICS; obsI++) {
    thislik = 0.0;
    
    for (int intI = 0; intI < n_integrationPointsUrbanMICS; intI++) {
      thisIndex = num_iUrbanMICS * intI + obsI;
      
      Type eta = alpha + fe_iUrbanMICS(thisIndex) + u_spatial(areaidxlocUrbanMICS(obsI)) + nuggetUrbMICS(obsI);
      
      thisWeight = wUrbanMICS(obsI, intI);
      
      if(thisWeight > 0.0) {
        thislik += thisWeight * dbinom_robust(y_iUrbanMICS(obsI), n_iUrbanMICS(obsI), eta, false);
      }
    }
    
    if(thislik > 0.0) {
      jnll -= log(thislik);
    }
  }
  
  // Rural MICS
  for (int obsI = 0; obsI < num_iRuralMICS; obsI++) {
    thislik = 0.0;
    
    for (int intI = 0; intI < n_integrationPointsRuralMICS; intI++) {
      thisIndex = num_iRuralMICS * intI + obsI;
      
      Type eta = alpha + fe_iRuralMICS(thisIndex) + u_spatial(areaidxlocRuralMICS(obsI)) + nuggetRurMICS(obsI);
      
      thisWeight = wRuralMICS(obsI, intI);
      
      if(thisWeight > 0.0) {
        thislik += thisWeight * dbinom_robust(y_iRuralMICS(obsI), n_iRuralMICS(obsI), eta, false);
      }
    }
    
    if(thislik > 0.0) {
      jnll -= log(thislik);
    }
  }
  
  // Report
  REPORT(log_tau);
  REPORT(log_tauEps);
  REPORT(alpha);
  REPORT(beta);
  REPORT(u_spatial);
  REPORT(sigma_u);
  REPORT(sigmaEps);
  REPORT(nuggetUrbMICS);
  REPORT(nuggetRurMICS);
  
  return jnll;
}
